tsl/src/bgw_policy/job.c

```diff
@@ -513,24 +513,60 @@ policy_invoke_recompress_chunk(Chunk *chunk)
 bool
 policy_compression_execute(int32 job_id, Jsonb *config)
 {
-	int32 chunkid;
+	List *chunkid_lst;
+	ListCell *lc;
 	const Dimension *dim;
 	PolicyCompressionData policy_data;
+	bool distributed, used_portalcxt = false;
+	MemoryContext saved_cxt, multitxn_cxt;
 
 	policy_compression_read_and_validate_config(config, &policy_data);
 	dim = hyperspace_get_open_dimension(policy_data.hypertable->space, 0);
-	chunkid = get_chunk_to_compress(dim, config);
+	distributed = hypertable_is_distributed(policy_data.hypertable);
+	/* we want the chunk id list to survive across transactions. So alloc in
+	 * a different context
+	 */
+	if (PortalContext)
+	{
+		/*if we have a portal context use that - it will get freed automatically*/
+		multitxn_cxt = PortalContext;
+		used_portalcxt = true;
+	}
+	else
+	{
+		/* background worker job does not go via usual CALL path, so we do
+		 * not have a PortalContext */
+		multitxn_cxt =
+			AllocSetContextCreate(TopMemoryContext, "CompressionJobCxt", ALLOCSET_DEFAULT_SIZES);
+	}
+	saved_cxt = MemoryContextSwitchTo(multitxn_cxt);
+	chunkid_lst = get_chunk_to_compress(dim, config);
+	MemoryContextSwitchTo(saved_cxt);
 
-	if (chunkid == INVALID_CHUNK_ID)
+	if (!chunkid_lst)
+	{
 		elog(NOTICE,
 			 "no chunks for hypertable %s.%s that satisfy compress chunk policy",
 			 policy_data.hypertable->fd.schema_name.data,
 			 policy_data.hypertable->fd.table_name.data);
-
-	if (chunkid != INVALID_CHUNK_ID)
+		ts_cache_release(policy_data.hcache);
+		if (!used_portalcxt)
+			MemoryContextDelete(multitxn_cxt);
+		return true;
+	}
+	ts_cache_release(policy_data.hcache);
+	if (ActiveSnapshotSet())
+		PopActiveSnapshot();
```

Comment:
won't this unbalance the snapshot stack if we don't keep track of this

Reply:
If we get here , we have atleast 1 chunk that needs processing. The loop commits the current transaction (line 570). You have to pop any active snapshots before the commit txn.  We start a new txn at 571. So we don't need to keep track of the snapshots from previous txns.

Reply:
maybe add this as comment

Reply:
You should probably add a comment about this.

Reply:
added.